### ISSUE TITLE
Variable input sequence length

### ISSUE THREAD

**Issue by ruoyxue** [state-spaces/mamba#236]
Dear authors,
This is an amazing work! 
I'm working with variable sequence lengths of video data. In one batch, there could be several videos with different frame numbers, and they will be padded to the same length. When I use transformer, I use attention masks to solve the problem of variable input lengths, but I do not see a similar mask in the Mamba forward function. Is there any solutions for dealing with variable lengths in a batch when using Mamba? Thanks!

**Comment by ruoyxue** [comment]
What's more, I'm using bidirectional Mamba, is there any solutions for processing data with variable lengths in a batch?

**Comment by tridao** [comment]
Variable length is not currently implemented but will be in the future. For now you can pad your sequences.

**Comment by EricPaul03** [comment]
I want to know what actions I should take on my sequence to achieve the entire pad process. For example, if my sequence length is 128 and I want to pad to a length of 200, can I directly cut it back to the first 128 before the skip connection of Mamba? The output of the last 72 dimensions will not interfere with my results, right?

**Comment by tridao** [comment]
Yes padding tokens should be on the right.

**Comment by zigzagcai** [comment]
We have implemented variable-length support for mamba block. https://github.com/state-spaces/mamba/pull/244

**Comment by ZJEast** [comment]
any progress ? I want to use variable-length sequences, too. And test the performance of Mamba on some Reinforcement Learning tasks.

**Comment by ZJEast** [comment]
I think it might be an alternative solution to use "gather" to solve this problem. Like

```
    def hidden_state(self, input_ids: Tensor, input_len: Tensor):
        hs: Tensor = self.backbone(input_ids, inference_params=None)
        B, L, D = hs.shape
        l = (input_len - 1).view(B, 1, 1)
        hs = hs.gather(1, l.expand(B, 1, D))
        return hs.view(B, D)
```

Here "input_len" is the length of the sequence.

**Comment by EricPaul03** [comment]
> I think it might be an alternative solution to use "gather" to solve this problem. Like
> 
> ```
>     def hidden_state(self, input_ids: Tensor, input_len: Tensor):
>         hs: Tensor = self.backbone(input_ids, inference_params=None)
>         B, L, D = hs.shape
>         l = (input_len - 1).view(B, 1, 1)
>         hs = hs.gather(1, l.expand(B, 1, D))
>         return hs.view(B, D)
> ```
> 
> Here "input_len" is the length of the sequence.

Hello, thank you very much for your answer. I don't quite understand your code yet. Can you tell me where to put this code and how to understand it? Thank you again!

**Comment by ZJEast** [comment]
As long as Mamba is a RNN architecture, the output of the padding words is dirty, what you need to do is just to drop the dirty output and prevent them from inference. I think it can be done in many various ways in pytorch. The padding should be on the right.

**Comment by EricPaul03** [comment]
> Because Mamba is a RNN architecture, the output of the padding words is dirty, what you need to do is just to drop the dirty output and prevent them from inference. I think it can be done in many various ways in pytorch. The padding should be on the right.
> 
> It really works for me.

But for performance reasons, I would like to use bidirectional mamba, which includes two scanning processes from left to right and from right to left. Is it not possible to directly fill on the right side?

**Comment by ZJEast** [comment]
[1, 2, 3] -> [1, 2, 3, 3, 2, 1]  ???
maybe you can try this preprocessing.
